src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -399,31 +408,86 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
       reporter.cancelled || unit.isJava && this.id > maxJavaPhase
     }
 
-    final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
-      if ((unit ne null) && unit.exists)
-        lastSeenSourceFile = unit.source
+    // Cleanup method needed for some phases, for example typer
+    def afterUnit(unit: CompilationUnit): Unit = {}
 
-      if (settings.debug && (settings.verbose || currentRun.size < 5))
-        inform("[running phase " + name + " on " + unit + "]")
-      if (!cancelled(unit)) {
-        currentRun.informUnitStarting(this, unit)
-        try withCurrentUnitNoLog(unit)(task)
-        finally currentRun.advanceUnit()
+    def run(): Unit = {
+      assertOnMain()
+
+      if (isDebugPrintEnabled) inform("[running phase " + name + " on " + currentRun.size +  " compilation units]")
+
+      implicit val ec: ExecutionContextExecutor = createExecutionContext()
```

Comment:
I wonder why can't reuse the same executionContext across all phases - to create a treadpull is not a cheap operation IIRC

Reply:
Good point, I don't see why we cant

Reply:
no we cant - its wiring requires that we generate for each phase. The reuse or otherwise has to be managed by the async helper to get the instrumentation